File: src-tauri/src/models/draft.rs
use serde::{Deserialize, Serialize};
use crate::models::{PlayerTag, Position};

/// 选秀球员 (选秀池)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DraftPlayer {
    pub id: u64,
    pub save_id: String,
    pub season_id: u64,
    pub region_id: u64,
    /// 选秀排名 (1=状元, 2=榜眼, 3=探花...)
    pub draft_rank: u8,
    pub game_id: String,
    pub real_name: Option<String>,
    pub nationality: Option<String>,
    pub age: u8,
    pub ability: u8,
    pub potential: u8,
    pub tag: PlayerTag,
    pub position: Option<Position>,
    /// 是否已被选中
    pub is_picked: bool,
    /// 被哪支队伍选中
    pub picked_by_team_id: Option<u64>,
}

/// 选秀顺位
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DraftOrder {
    pub id: u64,
    pub save_id: String,
    pub season_id: u64,
    pub region_id: u64,
    pub team_id: u64,
    /// 夏季赛排名
    pub summer_rank: u32,
    /// 选秀顺位 (抽签后)
    pub draft_position: u32,
    /// 抽签结果描述
    pub lottery_result: Option<String>,
}

/// 选秀结果
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DraftResult {
    pub id: u64,
    pub save_id: String,
    pub season_id: u64,
    pub region_id: u64,
    pub draft_player_id: u64,
    pub team_id: u64,
    /// 第几顺位
    pub pick_number: u32,
    /// 创建的选手ID
    pub player_id: Option<u64>,
}

/// 选秀抽签概率配置
/// 排名越靠后，获得高顺位的概率越高
pub fn draft_lottery_weights() -> Vec<(u32, Vec<f64>)> {
    // (排名, [获得1号签概率, 获得2号签概率, ...])
    // 排名1-14，每个排名都有一个概率分布
    vec![
        // 第14名 (最高获得状元签概率)
        (14, vec![0.25, 0.20, 0.15, 0.10, 0.08, 0.06, 0.05, 0.04, 0.03, 0.02, 0.01, 0.005, 0.003, 0.002]),
        // 第13名
        (13, vec![0.20, 0.22, 0.15, 0.10, 0.08, 0.07, 0.06, 0.05, 0.03, 0.02, 0.01, 0.007, 0.003]),
        // 第12名
        (12, vec![0.15, 0.18, 0.18, 0.12, 0.10, 0.08, 0.06, 0.05, 0.04, 0.02, 0.015, 0.005]),
        // 第11名
        (11, vec![0.12, 0.15, 0.16, 0.15, 0.12, 0.10, 0.08, 0.05, 0.04, 0.02, 0.01]),
        // 第10名
        (10, vec![0.10, 0.10, 0.14, 0.15, 0.15, 0.12, 0.10, 0.06, 0.04, 0.04]),
        // 第9名
        (9, vec![0.08, 0.08, 0.10, 0.12, 0.15, 0.15, 0.12, 0.10, 0.06, 0.04]),
        // 第8名
        (8, vec![0.05, 0.04, 0.06, 0.10, 0.12, 0.15, 0.16, 0.15, 0.10, 0.07]),
        // 第7名
        (7, vec![0.03, 0.02, 0.04, 0.08, 0.10, 0.13, 0.15, 0.18, 0.15, 0.12]),
        // 第6名
        (6, vec![0.015, 0.01, 0.015, 0.05, 0.08, 0.10, 0.14, 0.17, 0.20, 0.175]),
        // 第5名
        (5, vec![0.005, 0.005, 0.01, 0.02, 0.05, 0.07, 0.10, 0.15, 0.22, 0.27]),
        // 第1-4名通常不参与选秀抽签，保持原顺位
        (4, vec![0.0; 14]),
        (3, vec![0.0; 14]),
        (2, vec![0.0; 14]),
        (1, vec![0.0; 14]),
    ]
}

File: src/api/draft.ts
import { invokeCommand } from './client'

// ========================================
// Draft System
// ========================================

export interface DraftPlayer {
  id: number
  game_id: string
  position: string
  nationality: string | null
  age: number
  ability: number
  potential: number
  tag: string
  draft_rank: number
  is_picked: boolean
}

export interface DraftOrder {
  team_id: number
  team_name: string
  original_team_id: number | null
  summer_rank: number
  draft_position: number
  lottery_result: string | null
}

export interface DraftPick {
  pick_number: number
  team_id: number
  team_name: string
  player: DraftPlayer
}

export interface DraftResultItem {
  pick_number: number
  team_id: number
  team_name: string
  player_id: number
  player_name: string
  position: string
  ability: number
  potential: number
}

export interface DraftRegionStatus {
  status: 'not_started' | 'roster_drawn' | 'lottery_done' | 'completed'
  draft_players: DraftPlayer[]
  draft_results: DraftResultItem[]
  draft_orders: DraftOrder[]
  total_players: number
  picked_count: number
}

export interface DraftPoolPlayer {
  id: number
  game_id: string
  real_name: string | null
  nationality: string | null
  age: number
  ability: number
  potential: number
  position: string
  tag: string
  status: string
}

export interface NewDraftPoolPlayer {
  game_id: string
  real_name?: string | null
  nationality?: string | null
  age: number
  ability: number
  potential: number
  position: string
  tag: string
}

export interface UpdateDraftPoolPlayer {
  gameId: string
  ability: number
  potential: number
  position: string
  tag: string
}

export const draftApi = {
  generateDraftPool: (regionId: number, poolSize?: number) =>
    invokeCommand<DraftPlayer[]>('generate_draft_pool', {
      regionId,
      poolSize: poolSize ?? 14
    }),

  runDraftLottery: (regionId: number) =>
    invokeCommand<DraftOrder[]>('run_draft_lottery', { regionId }),

  getDraftOrder: (regionId: number) =>
    invokeCommand<DraftOrder[]>('get_draft_order', { regionId }),

  getAvailableDraftPlayers: (regionId: number) =>
    invokeCommand<DraftPlayer[]>('get_available_draft_players', { regionId }),

  makeDraftPick: (regionId: number, teamId: number, playerId: number) =>
    invokeCommand<DraftPick>('make_draft_pick', { regionId, teamId, playerId }),

  aiAutoDraft: (regionId: number) =>
    invokeCommand<DraftPick[]>('ai_auto_draft', { regionId }),

  // 选手池管理
  getDraftPoolPlayers: (regionId: number) =>
    invokeCommand<DraftPoolPlayer[]>('get_draft_pool_players', { regionId }),

  addDraftPoolPlayers: (regionId: number, players: NewDraftPoolPlayer[]) =>
    invokeCommand<number>('add_draft_pool_players', { regionId, players }),

  updateDraftPoolPlayer: (playerId: number, data: UpdateDraftPoolPlayer) =>
    invokeCommand<void>('update_draft_pool_player', { playerId, ...data }),

  deleteDraftPoolPlayers: (regionId: number, playerIds?: number[]) =>
    invokeCommand<number>('delete_draft_pool_players', { regionId, playerIds }),

  generateRookies: (regionId: number, count?: number, seed?: number, abilityMin?: number, abilityMax?: number) =>
    invokeCommand<DraftPoolPlayer[]>('generate_rookies', { regionId, count, seed, abilityMin, abilityMax }),

  // 选秀状态查询
  getDraftRegionStatus: (regionId: number, seasonId?: number) =>
    invokeCommand<DraftRegionStatus>('get_draft_region_status', { regionId, seasonId }),
}
